src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -716,26 +711,45 @@ public boolean canResume() {
         this.owner = owner;
+
         try {
-            initializeStorage();
             try {
-                if (!isComplete()) {
-                    if (canResume()) {
-                        loadProgramAsync(getProgramDataFile());
-                    } else {
-                        // TODO if possible, consider trying to close out unterminated blocks to keep existing graph history
-                        // That way we can visualize the graph in some error cases.
-                        LOGGER.log(Level.WARNING, "Pipeline state not properly persisted, cannot resume "+owner.getUrl());
-                        throw new IOException("Cannot resume build -- was not cleanly saved when Jenkins shut down.");
-                    }
-                } else if (done && !super.isComplete()) {
+                initializeStorage();  // Throws exception and bombs out if we can't load FlowNodes
+            } catch (Exception ex) {
+                LOGGER.log(Level.WARNING, "Error initializing storage and loading nodes, will try to create placeholders for: "+this, ex);
+                createPlaceholderNodes(ex);
+                return;
+            }
+        } catch (Exception ex) {
+            done = true;
+            programPromise = Futures.immediateFailedFuture(ex);
+            throw new IOException("Failed to even create placeholder nodes for execution", ex);
+        }
+
+        try {
+            if (isComplete()) {
+                if (done == Boolean.TRUE && !super.isComplete()) {
                     LOGGER.log(Level.WARNING, "Completed flow without FlowEndNode: "+this+" heads:"+getHeadsAsString());
                 }
-            } catch (Exception e) {  // Multicatch ensures that failure to load does not nuke the master
-                SettableFuture<CpsThreadGroup> p = SettableFuture.create();
-                programPromise = p;
-                loadProgramFailed(e, p);
+                if (super.isComplete() && done != Boolean.TRUE) {
+                    LOGGER.log(Level.WARNING, "Flow has FlowEndNode, but is not marked as done, fixing this for"+this);
+                    done = true;
+                    saveOwner();
+                }
+            } else {  // See if we can/should resume build
+                if (canResume()) {
+                    loadProgramAsync(getProgramDataFile());
+                } else {
+                    // TODO if possible, consider trying to close out unterminated blocks to keep existing graph history
+                    // That way we can visualize the graph in some error cases.
+                    LOGGER.log(Level.WARNING, "Pipeline state not properly persisted, cannot resume "+owner.getUrl());
+                    throw new IOException("Cannot resume build -- was not cleanly saved when Jenkins shut down.");
+                }
             }
+        } catch (Exception e) {  // Multicatch ensures that failure to load does not nuke the master
```

Comment:
BTW this is not technically a multicatch.

Reply:
Yeah, I need to update that comment since it's just a generic Exception now, good point.